lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -817,6 +817,64 @@ namespace Js
         return isPropertyDescriptorDefined;
     }
 
+    Var JavascriptObject::EntryGetOwnPropertyDescriptors(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+
+        ARGUMENTS(args, callInfo);
+        ScriptContext* scriptContext = function->GetScriptContext();
+
+        Assert(!(callInfo.Flags & CallFlags_New));
+
+        RecyclableObject* obj = nullptr;
+
+
+        if (args.Info.Count < 2)
+        {
+            obj = RecyclableObject::FromVar(JavascriptOperators::ToObject(scriptContext->GetLibrary()->GetUndefined(), scriptContext));
+        }
+        else
+        {
+            // Convert the argument to object first
+            obj = RecyclableObject::FromVar(JavascriptOperators::ToObject(args[1], scriptContext));
+        }
+
+        // If the object is HostDispatch try to invoke the operation remotely
+        if (obj->GetTypeId() == TypeIds_HostDispatch)
+        {
+            Var result;
+            if (obj->InvokeBuiltInOperationRemotely(EntryGetOwnPropertyDescriptors, args, &result))
+            {
+                return result;
+            }
+        }
+
+        Var ownPropertyKeys = JavascriptOperators::GetOwnPropertyKeys(obj, scriptContext);
+        JavascriptArray *ownPropsArray = JavascriptArray::FromVar(ownPropertyKeys);
+
+        RecyclableObject* resultObj = scriptContext->GetLibrary()->CreateObject(true);
```

Comment:
`requestedSlots` is correct.  This means we will allocate the object with its header and its initial slots in one allocation.  If later more properties are added to it they will move to an auxiliary slot array.  Hmm, that could be bad if there are a lot of properties.  Let me think about this more.  It will be fine to take this as is and change it later if this is a problem.

The type is a lot more complicated.  A type is the mapping of property names to property slot indexes in the object's slot array.  Every object has a pointer to its type in its header data.  However, it is non-trivial to know if a type can be shared.  Deep knowledge of the type system would be best to decide how to handle this.  I'll ask around.  I think it would be fine to not bother with the type, but if it is safe and cheap to do it would be best to share.